### ISSUE TITLE
Top-K Map Match Result Differentiation

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#970]
So lets start with a picture:

![k](https://user-images.githubusercontent.com/697548/31675847-ccdaef5c-b333-11e7-8a6d-4f6da7bb3229.png)

You'll see a hypothetical 3 point input path. Each red input point with its radius of accuracy will have some number of candidates. In this case the first point has 1 candidate, the second has 3 candidates (2 are overlapping), and the last point has 2 candidates.

You'll notice in the results that `k=0` and `k=1` only differ in that for the second point they either use the point along the first edge or the point at the start of the second edge. Otherwise the paths are the same. The score for the `k=1` will indeed be less than that of `k=0`. Similarly we get a result for `k=2` and `k=3` where the paths are essentially the same just that one uses the candidate along the edge where the other uses the candidate at the beginning of the next edge.

This makes sense mathematically given the positional inaccuracy of the input but actually including the result for both `k=1` and `k=3` doesn't really help the client. Specifically I'm wondering if the client wants to the top-k results where the paths are actually different not just the probability. So I'm just wondering what others opinions are on this.

**Comment by ptpt** [comment]
If the client only cares about the latest candidate, it makes sense to skip all next paths until you get a path in which the latest candidate ends at a different location (e.g. k=2).

**Comment by kevinkreiser** [comment]
it was my opinion that the client might rather know if the path were different at all, even if it did end at the same candidate. in this case different path means different list of edges.

**Comment by ptpt** [comment]
Then we can skip next paths until we get a path with a different list of edges.

**Comment by kevinkreiser** [comment]
yeah that is what i was wondering, do you think this is desirable?

**Comment by ptpt** [comment]
Yes, it's probably the only choice. My concern is the performance but I think it should be okay with small k and a small number of candidates.

**Comment by noblige** [comment]
@kevinkreiser I'd like alternative paths to be really alternative. I seem to be running into this issue pretty frequently (both main and alternative paths are the same).

**Comment by kevinkreiser** [comment]
ok sounds good ill take a crack at this, it seems straight forward to me.

**Comment by kevinkreiser** [comment]
Ok ive got an implementation for this that I'm currenlty testing but it occurs to me there are some edge cases. Specifically, are these unique or not:

![k1](https://user-images.githubusercontent.com/697548/31822856-1cb029ca-b578-11e7-9212-1c8afbda550a.png)

Technically the path in `k=0` has only 1 edge and the path in `k=1` has 2 edges, where the first edge isn't used at all because the correlation point is the end of the edge, none the less the edge is included in the path. @ptpt @noblige should `k=1` be thrown out?

**Comment by noblige** [comment]
I think it should be thrown out. I can hardly think of a use case where you would want both as candidates.

**Comment by kevinkreiser** [comment]
was thinking same

**Comment by ptpt** [comment]
Regarding the performance, there is a notable optimization to me.

Assume that in the figure below, there are only two distinct paths: the red path and the green path.

After we get the red path (the best one), we can remove candidates which are impossible to be used to construct distinct paths from the red path. For example, the candidate 1 and candidate 2 can be removed. Because their paths to the candidates at next time (3, 4, 5, 6) can't make any difference from the path starting from candidate 0 (via 4TH ST). By removing these "obviously impossible" candidates, we should be able to get the green path in a few iterations or just next iteration.

This is a rough idea. We need a clear definition of if two paths are distinct or equal first, then we know which candidates to remove.

![intro](https://user-images.githubusercontent.com/87357/31930100-c5a78770-b89e-11e7-8ac1-eefa93c11f09.png)

**Comment by kevinkreiser** [comment]
The idea of distinctness I was using was that if one path was found wholly within another path (with a caveat that when there is node snapping on the end or beginning of a result you can ignore those edges) then the two are not unique.

I had been thinking the same thing about removing candidates from the columns that wouldnt make any difference in the paths but i couldnt reason how to prove a taking out a given candidate would do that without removing future distinct results.

@ptpt do you have any thoughts about how to be sure removing a given candidate is safe?

**Comment by kevinkreiser** [comment]
Ok so we compute the red path as:

    time 0 = 0 -> 4
    time 1 = 4 -> 9
    time 2 = 9 -> 11

Then when we remove that path we could also remove some candidate pairs that have redundancies with the red route:

    time = 0 has candidates 0,1,2
    paths from 0 to 3,4,5,6 contain or are contained in the paths from 1 to 3,4,5,6, so remove 1
    paths from 0 to 3,4,5,6 contain or are contained in the paths from 2 to 3,4,5,6, so remove 2
    time = 1 has candidates 3,4,5,6
    paths from 4 to 7,8,9,10 contain or are contained in the paths from 3 to 7,8,9,10, so remove 3
    time = 2 has candidates 7,8,910
    paths from 9 to 11,12 contain or are contained in the paths from 7 to 11,12, so remove 7
    paths from 9 to 11,12 contain or are contained in the paths from 8 to 11,12, so remove 8
    paths from 9 to 11,12 contain or are contained in the paths from 10 to 11,12, so remove 10

Then we run again and find the green path as the next most optimal:

    time 0 = 0 -> 6
    time 1 = 6 -> 9
    time 2 = 9 -> 11

Then when we remove that path we could also remove some candidate pairs that have redundancies with the green route:

    time = 0 has candidate 0
    no more candidates to remove
    time = 1 has candidates 4,5,6
    paths from 6 to 7,8,9,10 contain or are contained in the paths from 5 to 7,8,9,10, so remove 5
    time = 2 has candidates 9
    no more candidates to remove

And if we ran again i suppose we'd get the path that looked like the red path but ended at candidate 12, at which point we'd remove candidate 11 maybe? I cant tell if we can remove candidates from the last time slot. Does this algorithm seem correct, I'm still mulling over whether its logically sound or not.

**Comment by kevinkreiser** [comment]
@ptpt would you mind if we merge this with the hardcoded limits in the interim? i would like to promote our discussion about removing candidates that contain only redundant paths to a proper issue and work against it in a separate pr. is this ok with you?

**Comment by ptpt** [comment]
Sure @kevinkreiser!
